src/async_impl/decoder.rs

```diff
@@ -223,8 +223,9 @@ impl Stream for Gzip {
         };
 
         match read {
-            Ok(read) if read == 0 => {
-                Ok(Async::Ready(None))
+            Ok(read) if read == 0 => match self.inner.get_mut().stream.poll() {
+                Ok(Async::Ready(Some(_))) => Ok(Async::Ready(None)),
```

Comment:
Maybe should return an error here.

Reply:
This case would mean gzip thought it was done, but there was more response body anyways? Ya, sounds dangerous and probably should error...

Also, this could skip some of the buffered bytes in `ReadState`. Maybe just call `.get_mut().read(&mut [0])`?

Reply:
done